### DISCUSSION TITLE
Suricata Rule Payload Using Variable

### DISCUSSION THREAD

**Discussion by Mfisher925** [Security-Onion-Solutions/securityonion#6914] [Unsupported Versions] [answered]
I was wondering if it is possible to use a $VARIABLE to match for the payload **content** within a Suricata rule.  Specifically we are trying to tune this rule to exclude a list of common .cloud domains.  

alert dns $HOME_NET any -> any any (msg:"ET INFO Observed DNS Query to .cloud TLD"; dns.query; **content:".cloud";** nocase; endswith; reference:url,www.spamhaus.org/statistics/tlds/; classtype:bad-unknown; sid:90000000; rev:5; metadata:affected_product Any, attack_target Client_Endpoint, created_at 2019_08_13, deployment Perimeter, former_category INFO, signature_severity Major, updated_at 2020_09_17;)

alert dns $HOME_NET any -> any any (msg:"ET INFO Observed DNS Query to .cloud TLD"; dns.query; **content:".cloud"; content:!"alienvault.cloud"; content:!"sensor.cloud.tenable.com"**; nocase; endswith; reference:url,www.spamhaus.org/statistics/tlds/; classtype:bad-unknown; sid:90000000; rev:5; metadata:affected_product Any, attack_target Client_Endpoint, created_at 2019_08_13, deployment Perimeter, former_category INFO, signature_severity Major, updated_at 2020_09_17;)

**Comment by dougburks** [comment, marked as answer]
It looks like you've made your own local copy of SID 2027865 as SID 90000000. If it were me, I think I'd just keep the existing 2027865 and change the content match using these instructions:
https://docs.securityonion.net/en/2.3/managing-alerts.html#modify-the-sid

**Reply by Mfisher925** [reply]
Thanks Doug,

I tried the following, and it accepted the command but when I grep all.rules, the rule is unchanged and if I try to apply that command again console states "already exists in modify".

> [somgr@seconman nids]$ sudo so-rule modify add 2027865 "content:".cloud"" "content:".cloud"; content:\\!"alienvault.cloud.com""
[sudo] password for somgr:
Configuration updated. Would you like to apply your changes now? (y/N) y
Applying changes:
Syncing config files...
Updating rules...
2022-01-20 21:18:38,362 - <INFO> - Loading ./rulecat.conf.
2022-01-20 21:18:38,365 - <INFO> - Forcing Suricata version to 6.0.
2022-01-20 21:18:38,367 - <INFO> - Fetching https://rules.emergingthreats.net/open/suricata-6.0.0/emerging.rules.tar.gz.
 100% - 3144551/3144551
2022-01-20 21:18:38,905 - <INFO> - Done.
2022-01-20 21:18:39,085 - <INFO> - Ignoring file rules/emerging-deleted.rules
2022-01-20 21:18:39,086 - <INFO> - Loading local file /opt/so/rules/nids/local.rules
2022-01-20 21:18:43,944 - <INFO> - Loaded 31524 rules.
2022-01-20 21:18:44,470 - <INFO> - Disabled 2 rules.
2022-01-20 21:18:44,470 - <INFO> - Enabled 0 rules.
2022-01-20 21:18:44,470 - <INFO> - Modified 0 rules.
2022-01-20 21:18:44,470 - <INFO> - Dropped 0 rules.
2022-01-20 21:18:44,727 - <INFO> - Enabled 131 rules for flowbit dependencies.
2022-01-20 21:18:49,939 - <INFO> - Writing rules to /opt/so/rules/nids/all.rules: total: 31524; enabled: 24192; added: 0; removed 0; modified: 1
2022-01-20 21:18:50,311 - <INFO> - Done.
[somgr@seconman nids]$ grep 2027865 all.rules
alert dns $HOME_NET any -> any any (msg:"ET INFO Observed DNS Query to .cloud TLD"; dns.query; content:".cloud"; nocase; endswith; reference:url,www.spamhaus.org/statistics/tlds/; classtype:bad-unknown; sid:2027865; rev:4; metadata:affected_product Any, attack_target Client_Endpoint, created_at 2019_08_13, deployment Perimeter, former_category INFO, signature_severity Major, updated_at 2020_09_17;)
[somgr@seconman nids]$ sudo so-rule modify add 2027865 "content:".cloud"" "content:".cloud"; content:\!"alienvault.cloud.com""
2027865 "content:.cloud" "content:.cloud; content:\!alienvault.cloud.com" already exists in modify

**Reply by dougburks** [reply]
Since your content matches include quotes, I suspect the quotes would need to be escaped as shown in the documentation.

**Reply by Mfisher925** [reply]
Ah ok, I had tried that but I must have missed one or two and it didn't want to accept the command.  Also since I need to add "!" to the content for it to ignore that content, I have to escape that for BASH to accept the command.  Which leaves the "\\!" in the modification which I don't believe should be there.  

This command is accepted, and the modification shows up when I use "so-rule modify list", however after running the command the output shows "0 modified" and the rule is unchanged in all.rules.

> [somgr@seconman nids]$ sudo so-rule modify add 2027865 "content:\".cloud\"" "content:\".cloud\"; content:\\!\"alienvault.cloud.com\""
[sudo] password for somgr:
Configuration updated. Would you like to apply your changes now? (y/N) y
Applying changes:
Syncing config files...
Updating rules...
2022-01-21 20:42:12,308 - <INFO> - Loading ./rulecat.conf.
2022-01-21 20:42:12,316 - <INFO> - Forcing Suricata version to 6.0.
2022-01-21 20:42:12,318 - <INFO> - Fetching https://rules.emergingthreats.net/open/suricata-6.0.0/emerging.rules.tar.gz.
 100% - 3145197/3145197
2022-01-21 20:42:13,918 - <INFO> - Done.
2022-01-21 20:42:14,334 - <INFO> - Ignoring file rules/emerging-deleted.rules
2022-01-21 20:42:14,335 - <INFO> - Loading local file /opt/so/rules/nids/local.rules
2022-01-21 20:42:22,425 - <INFO> - Loaded 31531 rules.
2022-01-21 20:42:22,930 - <INFO> - Disabled 3 rules.
2022-01-21 20:42:22,930 - <INFO> - Enabled 0 rules.
2022-01-21 20:42:22,930 - <INFO> - Modified 0 rules.
2022-01-21 20:42:22,930 - <INFO> - Dropped 0 rules.
2022-01-21 20:42:23,402 - <INFO> - Enabled 131 rules for flowbit dependencies.
2022-01-21 20:42:32,317 - <INFO> - Writing rules to /opt/so/rules/nids/all.rules: total: 31531; enabled: 24198; added: 0; removed 0; modified: 0
2022-01-21 20:42:32,954 - <INFO> - Done.
[somgr@seconman nids]$ sudo so-rule modify list
Modified rules + modifications:
>   2027865 "content:".cloud"" "content:".cloud"; content:\\!"alienvault.cloud.com""
[somgr@seconman nids]$ grep 2027865 all.rules
alert dns $HOME_NET any -> any any (msg:"ET INFO Observed DNS Query to .cloud TLD"; dns.query; content:".cloud"; nocase; endswith; reference:url,www.spamhaus.org/statistics/tlds/; classtype:bad-unknown; sid:2027865; rev:4; metadata:affected_product Any, attack_target Client_Endpoint, created_at 2019_08_13, deployment Perimeter, former_category INFO, signature_severity Major, updated_at 2020_09_17;)
[somgr@seconman nids]$

**Reply by paulgear** [reply]
> It looks like you've made your own local copy of SID 2027865 as SID 90000000. If it were me, I think I'd just keep the existing 2027865 and change the content match using these instructions: https://docs.securityonion.net/en/2.3/managing-alerts.html#modify-the-sid

Is modifying existing rules to add exceptions like this considered best practice for rule management?  This is something that has mystified me for a while.  I would have thought it desirable to leave existing rules unmodified as much as possible, and add bypass rules which override them.    It becomes unwieldy rather quickly when you have long lists of excluded domains (e.g. in relation to this particular rule, if you work with a lot of cloud companies, `.cloud` shows up as a DNS TLD quite often).  I can't help but feel there must be a better way.  Are there any tools to facilitate a smooth workflow for managing rules like that?

**Reply by dougburks** [reply]
> Is modifying existing rules to add exceptions like this considered best practice for rule management?

I prefaced my response with `If it were me` because there are multiple methods of achieving the same objective and different folks have different preferences. Use what works best for you.

> Are there any tools to facilitate a smooth workflow for managing rules like that?

We have plans to build an interface in SOC for managing rules.

**Reply by paulgear** [reply]
@dougburks Thanks - I'm looking forward to that SOC feature for sure.  In the meantime, is modifying existing rules and adding exceptions the best way to achieve the desired result of allow-listing DNS names?  One of the issues I've run into recently is that there are often 2 or 3 rules which trigger on the same site: there's a DNS lookup, then a TLS SNI hostname request, then a TLS SNI hostname response.  Is the best currently-available method to copy the site allow list into all of those rules?

**Reply by dougburks** [reply]
Here's another way to think about rules like this:

alert dns $HOME_NET any -> any any (msg:"ET INFO Observed DNS Query to .cloud TLD"; dns.query; content:".cloud"; nocase; endswith; reference:url,www.spamhaus.org/statistics/tlds/; classtype:bad-unknown; sid:2027865; rev:5; metadata:affected_product Any, attack_target Client_Endpoint, created_at 2019_08_13, deployment Perimeter, former_category INFO, signature_severity Major, updated_at 2020_09_17;)

First, note that the rule is in the `INFO` category and some organizations disable that category altogether.

For those organizations that don't disable the entire `INFO` category, they might look at that rule and determine that their organization expects to see legitimate traffic to cloud domains and thus disable that particular rule altogether.

They might then decide to write more focused rules looking for specific malicious cloud sites. Or they might decide to craft a Hunt query to look for cloud domains and filter out known good usage. Maybe even leverage ElastAlert or Playbook.

Just trying to share some additional options that other organizations have implemented. Use what works best for you and your organization.
